### PR TITLE
feature(appliance): UI installs Sourcegraph again

### PR DESCRIPTION
**Pull request by jdpleiness**
The appliance UI creates a `configMap` that triggers our reconciler to install Sourcegraph.

- Refactored React app to remove "demo" code where possible
- Added `status` functionality to get the status of deployments, statefulsets, or persistent volume claims.
- Fix small code issue where `FromInt` was deprecated
- Add reconciliation loop for the appliance backing `configMap`

Install:

```shell
❯ k get pods
NAME                                         READY   STATUS    RESTARTS      AGE
blobstore-568d57d477-k5vrh                   1/1     Running   0             15m
codeinsights-db-0                            2/2     Running   0             15m
codeintel-db-0                               2/2     Running   0             15m
gitserver-0                                  1/1     Running   1 (14m ago)   14m
grafana-0                                    1/1     Running   0             15m
indexed-search-0                             2/2     Running   0             15m
pgsql-0                                      2/2     Running   0             15m
precise-code-intel-worker-6fd4d6c8d5-4rjbc   1/1     Running   1 (14m ago)   15m
precise-code-intel-worker-6fd4d6c8d5-6fjdw   1/1     Running   1 (14m ago)   15m
prometheus-6cd94d7485-4cx7n                  1/1     Running   0             15m
redis-cache-f4dc7d6b8-2tp5v                  2/2     Running   0             15m
redis-store-5d6fcc9c84-7sj8v                 2/2     Running   0             15m
repo-updater-9f695b7d5-r28t7                 1/1     Running   5 (14m ago)   15m
searcher-0                                   1/1     Running   0             15m
sourcegraph-frontend-64cc4458cd-4vdwq        1/1     Running   0             15m
sourcegraph-frontend-64cc4458cd-gd7bf        1/1     Running   0             15m
symbols-0                                    1/1     Running   0             13m
syntect-server-6d5d55fb4f-tgbsc              1/1     Running   0             15m
worker-66b4cd79b5-zw844                      1/1     Running   1 (13m ago)   15m

```

## Test plan

Unit tests where applicable and tested locally via UI and local cluster

### CODE REVIEW

File: internal/appliance/json.go [hunk 1]
```diff
@@ -105,27 +108,60 @@ func (a *Appliance) getStageJSONHandler() http.Handler {
 
 func (a *Appliance) getInstallProgressJSONHandler() http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		currentTasks, progress := calculateProgress(installTasks())
```
Comment (jdpleiness) [inline internal/appliance/json.go:111]: This is here so we don't get a null pointer ref in the frontend for now. Ultimately this needs to be generated/updated on every call.

File: internal/appliance/json.go [hunk 2]
```diff
@@ -105,27 +108,60 @@ func (a *Appliance) getStageJSONHandler() http.Handler {
...
 
+		installProgress := struct {
```
Comment (jdpleiness) [inline internal/appliance/json.go:113]: This is here so we don't get a null pointer ref in the frontend for now. Ultimately this needs to be generated/updated on every call.

File: internal/appliance/json.go [hunk 3]
```diff
@@ -105,27 +108,60 @@ func (a *Appliance) getStageJSONHandler() http.Handler {
...
 
 		if err := a.readJSON(w, r, &input); err != nil {
 			a.badRequestResponse(w, r, err)
 			return
 		}
+
+		a.sourcegraph.Spec.RequestedVersion = input.Data
+
+		//TODO(jdpleiness) check form for value if this should be set or not
+		a.sourcegraph.SetLocalDevMode()
```
Comment (jdpleiness) [inline internal/appliance/json.go:152]: We need to read in our db connection strings and other info like localDevMode here

File: internal/appliance/reconciler/frontend.go [hunk 4]
```diff
@@ -156,8 +156,8 @@ func (r *Reconciler) reconcileFrontendDeployment(ctx context.Context, sg *config
 	dep := deployment.NewDeployment("sourcegraph-frontend", sg.Namespace, sg.Spec.RequestedVersion)
 	dep.Spec.Replicas = &cfg.Replicas
 	dep.Spec.Strategy.RollingUpdate = &appsv1.RollingUpdateDeployment{
-		MaxSurge:       pointers.Ptr(intstr.FromInt(2)),
-		MaxUnavailable: pointers.Ptr(intstr.FromInt(0)),
+		MaxSurge:       pointers.Ptr(intstr.FromInt32(2)),
```
Comment (jdpleiness) [inline internal/appliance/reconciler/frontend.go:159]: Just sneaking in some quick fixes for deprecation warnings here

File: internal/appliance/routes.go [hunk 5]
```diff
@@ -10,14 +10,15 @@ func (a *Appliance) Routes() *mux.Router {
...
 
-	r.Handle("/api/operator/v1beta1/stage", a.getStageJSONHandler())
-	r.Handle("/api/operator/v1beta1/install/progress", a.getInstallProgressJSONHandler())
-	r.Handle("/api/operator/v1beta1/maintenance/status", a.getStageJSONHandler())
-	r.Handle("/api/operator/v1beta1/fake/stage", a.postStageJSONHandler())
+	// Maintenance API URIs
+	r.Handle("/api/v1/appliance/status", a.getStatusJSONHandler()).Methods("GET")
+	r.Handle("/api/v1/appliance/status", a.postStatusJSONHandler()).Methods("POST")
+	r.Handle("/api/v1/appliance/install/progress", a.getInstallProgressJSONHandler()).Methods("GET")
+	r.Handle("/api/v1/appliance/maintenance/status", a.getMaintenanceStatusHandler()).Methods("GET")
```
Comment (jdpleiness) [inline internal/appliance/routes.go:21]: Updated URIs to look a little cleaner

File: internal/appliance/status.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+			Started:     false,
+			Finished:    false,
+			Weight:      1,
+		},
+	}
+}
+
+// IsObjectReady checks if a k8s object is ready, with the definition of ready depending on the object type.
+// Supported resource types are StatefulSets, Deployments, and PersistentVolumeClaims.
+func IsObjectReady(obj client.Object) (bool, error) {
```
Comment (jdpleiness) [inline internal/appliance/status.go:70]: We can use this to get a good idea if our services are ready in order to update the progress bar in the UI. Still need to write up the loops to iterate over our services though

File: internal/appliance/json.go (hunk 2 above)
Comment (craigfurman) [reply internal/appliance/json.go:113]: Is this basically stubbing the progress bar, as a stepping stone? If so let's leave a comment saying that's what it does. I am totally on board with getting the steel thread working and punting a more-accurate progress bar into follow-on backlog issues.

File: internal/appliance/json.go (hunk 3 above)
Comment (craigfurman) [reply internal/appliance/json.go:152]: In yesterday's team meeting, IIRC we punted remote DBs past the very first release. Let's create a follow-on issue for that.

re: Dev mode, IMO this code makes it too easy to accidentally never configure resources for real users. What if we conditionalize this line behind an env var check (e.g. `APPLIANCE_DEV_MODE`), and create a follow on issue to restore the dev-mode form field?

**Review by craigfurman — COMMENTED** [review]
Amazing to see the installation going again! When I ran this locally, I could get it to install all resources, which eventually converged to healthy, but the progress bar never left 0% 🤔 

Not sure if just my local environment or a real bug, let's pair on this later. Maybe, the desired end state is that success page we talked about, with a big "I'm ready" button that causes the ingress to flip for the first time? Presumably this would be backed by a new status to represent "first install done, waiting for admin acknowledgement" (maybe "awaiting-admin-post-first-install")? I think something to that effect was proposed yesterday.

**Review by craigfurman — APPROVED** [review]
We split the step after the progress bar into a follow-on: https://linear.app/sourcegraph/issue/REL-291/wait-for-admin-ui-integrates-with-backend

🚀 

Great work!

File: internal/appliance/json.go (hunk 2 above)
Comment (jdpleiness) [reply internal/appliance/json.go:113]: Yeah just a stub for now. Based on the meeting yesterday we can either fully mock this or hopefully tie it in to work in a follow up.

**Pull request merged**